# **Project Name** - Vaccination Data Analysis and Visualization

##### **Project Type**    - EDA, SQL & Power BI
##### **Contribution**    - Individual

## **Project Summary:**
Vaccination Data Analysis and Visualization is a public health project leveraging Python for data cleaning and exploratory analysis, SQL for structured database storage, and Power BI for interactive dashboard reporting. Global vaccination and disease datasets are cleaned, normalized, and stored in a relational database to answer key analytic questions and support evidence-based policy-making. The project delivers a robust analytics framework where stakeholders can explore vaccination trends, identify gaps in coverage, and assess the effectiveness of vaccination programs using dynamic visualizations and actionable insights.

## **GitHub Link -**
https://github.com/gvdharun/Vaccination-Data-Analysis-and-Visualization.git

# **Problem Statement:**
Analyze global vaccination data to understand trends in vaccination coverage, disease incidence, and effectiveness. Data will be cleaned, and stored in a SQL database. Power BI will be used to connect to the SQL database and create interactive dashboards that provide insights on vaccination strategies and their impact on disease control.

# ***Let's Begin !***

## **1. Know your data**

Import Libraries

In [37]:
# Import necessary libraries
# Data manipulation library
import pandas as pd
import numpy as np

# Database library
import mysql.connector as mysql
import sqlite3 as sql

# Data visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

Loading dataset

In [2]:
coverage = pd.read_csv(r"Data\coverage-data.csv")
incidence = pd.read_csv(r"Data\incidence-rate-data.csv")
reported = pd.read_csv(r"Data\reported-cases-data.csv")
vaccine_intro = pd.read_csv(r"Data\vaccine-introduction-data.csv")
vaccine_schedule = pd.read_csv(r"Data\vaccine-schedule-data.csv")

Dataset View

In [3]:
coverage.head()

GROUP CODE   NAME  YEAR  ANTIGEN  \
0  COUNTRIES  ABW  Aruba  2023      BCG   
1  COUNTRIES  ABW  Aruba  2023      BCG   
2  COUNTRIES  ABW  Aruba  2023  DIPHCV4   
3  COUNTRIES  ABW  Aruba  2023  DIPHCV4   
4  COUNTRIES  ABW  Aruba  2023  DIPHCV5   

                                 ANTIGEN_DESCRIPTION COVERAGE_CATEGORY  \
0                                                BCG             ADMIN   
1                                                BCG          OFFICIAL   
2  Diphtheria-containing vaccine, 4th dose (1st b...             ADMIN   
3  Diphtheria-containing vaccine, 4th dose (1st b...          OFFICIAL   
4  Diphtheria-containing vaccine, 5th dose (2nd b...             ADMIN   

  COVERAGE_CATEGORY_DESCRIPTION  TARGET_NUMBER   DOSES  COVERAGE  
0       Administrative coverage            NaN     NaN       NaN  
1             Official coverage            NaN     NaN       NaN  
2       Administrative coverage         1044.0   945.0     90.52  
3             Official coverage            NaN     NaN     90.52  
4       Administrative coverage         1219.0  1008.0     82.69

In [4]:
incidence.head()

GROUP CODE   NAME  YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023              CRS   
1  COUNTRIES  ABW  Aruba  2023       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023          MEASLES   
4  COUNTRIES  ABW  Aruba  2023            MUMPS   

              DISEASE_DESCRIPTION                     DENOMINATOR  \
0     Congenital rubella syndrome          per 10,000 live births   
1                      Diphtheria  per 1,000,000 total population   
2  Invasive meningococcal disease  per 1,000,000 total population   
3                         Measles  per 1,000,000 total population   
4                           Mumps  per 1,000,000 total population   

   INCIDENCE_RATE  
0             0.0  
1             0.0  
2             9.3  
3             NaN  
4             0.0

In [5]:
reported.head()

GROUP CODE   NAME  YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023              CRS   
1  COUNTRIES  ABW  Aruba  2023       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023          MEASLES   
4  COUNTRIES  ABW  Aruba  2023            MUMPS   

              DISEASE_DESCRIPTION  CASES  
0     Congenital rubella syndrome    0.0  
1                      Diphtheria    0.0  
2  Invasive meningococcal disease    1.0  
3                         Measles    NaN  
4                           Mumps    0.0

In [6]:
vaccine_intro.head()

ISO_3_CODE  COUNTRYNAME WHO_REGION  YEAR  \
0        AFG  Afghanistan       EMRO  2023   
1        AFG  Afghanistan       EMRO  2023   
2        AFG  Afghanistan       EMRO  2023   
3        AFG  Afghanistan       EMRO  2023   
4        AFG  Afghanistan       EMRO  2023   

                                   DESCRIPTION INTRO  
0             aP (acellular pertussis) vaccine    No  
1                          Hepatitis A vaccine    No  
2                          Hepatitis B vaccine   Yes  
3                              HepB birth dose   Yes  
4  Hib (Haemophilus influenzae type B) vaccine   Yes

In [7]:
vaccine_schedule.head()

ISO_3_CODE COUNTRYNAME WHO_REGION  YEAR VACCINECODE  \
0        ABW       Aruba       AMRO  2023  DTAPHIBIPV   
1        ABW       Aruba       AMRO  2023  DTAPHIBIPV   
2        ABW       Aruba       AMRO  2023  DTAPHIBIPV   
3        ABW       Aruba       AMRO  2023  DTAPHIBIPV   
4        ABW       Aruba       AMRO  2023     DTAPIPV   

                VACCINE_DESCRIPTION  SCHEDULEROUNDS  TARGETPOP  \
0  DTaP-Hib-IPV (acellular) vaccine               1        NaN   
1  DTaP-Hib-IPV (acellular) vaccine               2        NaN   
2  DTaP-Hib-IPV (acellular) vaccine               3        NaN   
3  DTaP-Hib-IPV (acellular) vaccine               4    B_2YL_W   
4      DTaP-IPV (acellular) vaccine               5  B_CHILD_W   

  TARGETPOP_DESCRIPTION   GEOAREA AGEADMINISTERED SOURCECOMMENT  
0       General/routine  NATIONAL              M2           NaN  
1       General/routine  NATIONAL              M4           NaN  
2       General/routine  NATIONAL              M6           NaN  
3       General/routine  NATIONAL             M15           NaN  
4       General/routine  NATIONAL              Y4           NaN

Dataset Shape

In [8]:
print("coverage: ", coverage.shape)
print("incidence: ", incidence.shape)
print("reported: ", reported.shape)
print("vaccine_intro: ", vaccine_intro.shape)
print("vaccine_schedule: ", vaccine_schedule.shape)

coverage:  (399858, 11)
incidence:  (84945, 8)
reported:  (84869, 7)
vaccine_intro:  (138320, 6)
vaccine_schedule:  (8052, 12)


Dataset info

In [9]:
print("coverage info: ")
coverage.info()

print("\nincidence info: ")
incidence.info()

print("\nreported info: ")
reported.info()

print("\nvaccine_intro info: ")
vaccine_intro.info()

print("\nvaccine_schedule info: ")
vaccine_schedule.info()

coverage info: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 399858 entries, 0 to 399857
Data columns (total 11 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   GROUP                          399858 non-null  object 
 1   CODE                           399858 non-null  object 
 2   NAME                           398584 non-null  object 
 3   YEAR                           399858 non-null  int64  
 4   ANTIGEN                        399858 non-null  object 
 5   ANTIGEN_DESCRIPTION            399858 non-null  object 
 6   COVERAGE_CATEGORY              399858 non-null  object 
 7   COVERAGE_CATEGORY_DESCRIPTION  399858 non-null  object 
 8   TARGET_NUMBER                  79030 non-null   float64
 9   DOSES                          79327 non-null   float64
 10  COVERAGE                       230477 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 33.6+ MB

incidence info: 

Duplicates Vales:

In [10]:
print("Duplicate rows in coverage: ", coverage.duplicated().sum())
print("Duplicate rows in incidence: ", incidence.duplicated().sum())
print("Duplicate rows in reported: ", reported.duplicated().sum())
print("Duplicate rows in vaccine_intro: ", vaccine_intro.duplicated().sum())
print("Duplicate rows in vaccine_schedule: ", vaccine_schedule.duplicated().sum())

Duplicate rows in coverage:  0
Duplicate rows in incidence:  0
Duplicate rows in reported:  0
Duplicate rows in vaccine_intro:  0
Duplicate rows in vaccine_schedule:  0


Missing values / Null values:

In [11]:
print("Null values in coverage:\n", coverage.isnull().sum())
print("\nNull values in incidence:\n", incidence.isnull().sum())
print("\nNull values in reported:\n", reported.isnull().sum())
print("\nNull values in vaccine_intro:\n", vaccine_intro.isnull().sum())
print("\nNull values in vaccine_schedule:\n", vaccine_schedule.isnull().sum())

Null values in coverage:
 GROUP                                 0
CODE                                  0
NAME                               1274
YEAR                                  0
ANTIGEN                               0
ANTIGEN_DESCRIPTION                   0
COVERAGE_CATEGORY                     0
COVERAGE_CATEGORY_DESCRIPTION         0
TARGET_NUMBER                    320828
DOSES                            320531
COVERAGE                         169381
dtype: int64

Null values in incidence:
 GROUP                      0
CODE                       0
NAME                       0
YEAR                       0
DISEASE                    0
DISEASE_DESCRIPTION        0
DENOMINATOR                0
INCIDENCE_RATE         23361
dtype: int64

Null values in reported:
 GROUP                      0
CODE                       0
NAME                       0
YEAR                       0
DISEASE                    0
DISEASE_DESCRIPTION        0
CASES                  19399
dtype: int64

Null

## Dataset Overview

The vaccination project dataset consists of five major tables, each providing detailed information related to vaccination coverage, disease incidence, reported cases, vaccine introductions, and vaccination schedules across different countries and years. Below is an in-depth description of each table, including their structure, key variables, and data completeness.

**1. Coverage Data Table**

- **Size:** 399,858 rows, 11 columns
- **Description:**  
  This table records vaccination coverage statistics by country, antigen (vaccine), and year. It provides insight into the percentage of the target population vaccinated, the number of doses administered, and coverage types.
- **Key Columns:**
  - `GROUP`: Country or group categorization (non-null).
  - `CODE`: ISO Alpha-3 country code (non-null).
  - `NAME`: Country name (mostly non-null, some missing).
  - `YEAR`: Year of record (non-null).
  - `ANTIGEN` and `ANTIGEN_DESCRIPTION`: Vaccine identifiers and descriptions (non-null).
  - `COVERAGE_CATEGORY` and `COVERAGE_CATEGORY_DESCRIPTION`: Type and detailed description of coverage (non-null).
  - `TARGET_NUMBER`: Number of individuals targeted for vaccination (missing in many rows).
  - `DOSES`: Number of doses administered (missing in many rows).
  - `COVERAGE`: Percentage coverage of the target population; has considerable missing data (about 42% missing).

**2. Incidence Rate Table**

- **Size:** 84,945 rows, 8 columns
- **Description:**  
  This dataset contains incidence rates of various diseases by country and year. It captures how prevalent certain vaccine-preventable diseases are within populations.
- **Key Columns:**
  - `GROUP`: Country or group categorization (non-null).
  - `CODE`: ISO Alpha-3 country code (non-null).
  - `NAME`: Country name (non-null).
  - `YEAR`: Year of record (non-null).
  - `DISEASE` and `DISEASE_DESCRIPTION`: Disease identifiers and detailed descriptions (non-null).
  - `DENOMINATOR`: Population basis for incidence calculation (e.g., per live births) (non-null).
  - `INCIDENCE_RATE`: Disease incidence rate; substantial missing data (~28% missing).

**3. Reported Cases Table**

- **Size:** 84,869 rows, 7 columns
- **Description:**  
  This table tracks the actual reported number of cases for various diseases by country and year. It complements incidence rates by providing absolute case counts.
- **Key Columns:**
  - `GROUP`: Country or group categorization (non-null).
  - `CODE`: ISO Alpha-3 country code (non-null).
  - `NAME`: Country name (non-null).
  - `YEAR`: Year of record (non-null).
  - `DISEASE` and `DISEASE_DESCRIPTION`: Disease identifiers and descriptions (non-null).
  - `CASES`: Number of reported disease cases; notable missing data (~23% missing).

**4. Vaccine Introduction Table**

- **Size:** 138,320 rows, 6 columns
- **Description:**  
  This dataset documents which vaccines have been introduced in each country’s vaccination program, by year, along with the corresponding WHO region.
- **Key Columns:**
  - `ISO_3_CODE`: ISO Alpha-3 country code (non-null).
  - `COUNTRYNAME`: Name of the country (non-null).
  - `WHO_REGION`: WHO region classification (non-null).
  - `YEAR`: Year of record (non-null).
  - `DESCRIPTION`: Vaccine or vaccine type description (non-null).
  - `INTRO`: Indicator flagging vaccine introduction status (non-null).

**5. Vaccine Schedule Data Table**

- **Size:** 8,052 rows, 12 columns
- **Description:**  
  This table provides detailed vaccination schedules for each country, including timing, target population, and geographical area for vaccine administration.
- **Key Columns:**
  - `ISO_3_CODE`: ISO Alpha-3 country code (non-null).
  - `COUNTRYNAME`: Country name (non-null).
  - `WHO_REGION`: WHO region classification (non-null).
  - `YEAR`: Year of record (non-null).
  - `VACCINECODE` and `VACCINE_DESCRIPTION`: Vaccine identifiers and descriptions (non-null).
  - `SCHEDULEROUNDS`: The dose or round number in the schedule (non-null).
  - `TARGETPOP`: Target population group (partially missing).
  - `TARGETPOP_DESCRIPTION`: Description of the target group (non-null).
  - `GEOAREA`: Geographic area of administration (partially missing).
  - `AGEADMINISTERED`: Age or time of vaccine administration (partially missing).
  - `SOURCECOMMENT`: Additional commentary or notes (highly missing).

---

## Duplicate values

| Table             | Duplicate Rows |
|-------------------|----------------|
| Coverage          | 0              |
| Incidence         | 0              |
| Reported Cases    | 0              |
| Vaccine Introduction | 0           |
| Vaccine Schedule  | 0              |

*No duplicate rows found in any of the tables, indicating unique records across datasets.*

---

## Missing Values

### Coverage Table

| Column                      | Missing Values | Notes                                      |
|-----------------------------|----------------|--------------------------------------------|
| GROUP                       | 0              | No missing values                           |
| CODE                        | 0              | No missing values                           |
| NAME                        | 1,274          | Small portion missing country names        |
| YEAR                        | 0              | Complete                                   |
| ANTIGEN                     | 0              | Complete                                   |
| ANTIGEN_DESCRIPTION         | 0              | Complete                                   |
| COVERAGE_CATEGORY           | 0              | Complete                                   |
| COVERAGE_CATEGORY_DESCRIPTION| 0              | Complete                                   |
| TARGET_NUMBER               | 320,828        | Large missing portion (~80% missing)       |
| DOSES                       | 320,531        | Large missing portion (~80% missing)       |
| COVERAGE                    | 169,381        | Significant missing data (~42% missing)    |

### Incidence Table

| Column             | Missing Values | Notes                                         |
|--------------------|----------------|-----------------------------------------------|
| GROUP              | 0              | Complete                                      |
| CODE               | 0              | Complete                                      |
| NAME               | 0              | Complete                                      |
| YEAR               | 0              | Complete                                      |
| DISEASE            | 0              | Complete                                      |
| DISEASE_DESCRIPTION| 0              | Complete                                      |
| DENOMINATOR        | 0              | Complete                                      |
| INCIDENCE_RATE     | 23,361         | Substantial missing values (~27% missing)    |

### Reported Cases Table

| Column             | Missing Values | Notes                                         |
|--------------------|----------------|-----------------------------------------------|
| GROUP              | 0              | Complete                                      |
| CODE               | 0              | Complete                                      |
| NAME               | 0              | Complete                                      |
| YEAR               | 0              | Complete                                      |
| DISEASE            | 0              | Complete                                      |
| DISEASE_DESCRIPTION| 0              | Complete                                      |
| CASES              | 19,399         | Notable missing values (~23% missing)        |

### Vaccine Introduction Table

| Column      | Missing Values | Notes                  |
|-------------|----------------|------------------------|
| ISO_3_CODE  | 0              | Complete               |
| COUNTRYNAME | 0              | Complete               |
| WHO_REGION  | 0              | Complete               |
| YEAR        | 0              | Complete               |
| DESCRIPTION | 0              | Complete               |
| INTRO       | 0              | Complete               |

### Vaccine Schedule Table

| Column                 | Missing Values | Notes                                      |
|------------------------|----------------|--------------------------------------------|
| ISO_3_CODE             | 0              | Complete                                   |
| COUNTRYNAME            | 0              | Complete                                   |
| WHO_REGION             | 0              | Complete                                   |
| YEAR                   | 0              | Complete                                   |
| VACCINECODE            | 0              | Complete                                   |
| VACCINE_DESCRIPTION    | 0              | Complete                                   |
| SCHEDULEROUNDS         | 0              | Complete                                   |
| TARGETPOP              | 4,257          | Moderate missing data (~53%)                |
| TARGETPOP_DESCRIPTION  | 0              | Complete                                   |
| GEOAREA                | 30             | Minor missing data                          |
| AGEADMINISTERED        | 1,045          | Notable missing values                      |
| SOURCECOMMENT          | 2,913          | Significant missing data (~36%)             |

---

*Overall, the datasets have no duplicate entries but contain varying degrees of missing data, especially in quantitative fields such as coverage targets, doses administered, incidence rates, and case counts that will require careful handling during data cleaning and analysis.*

---

## **2. Data Preprocessing**

Handling missing values:

In [12]:
# 1. Coverage data cleaning
coverage['NAME'] = coverage['NAME'].ffill()  # Fill missing values with the previous value
coverage = coverage.dropna(subset=['TARGET_NUMBER'])  # Drop rows with missing target values
coverage = coverage.dropna(subset=['DOSES'])  # Drop rows with missing doses values
coverage = coverage.dropna(subset=['COVERAGE'])  # Drop rows with missing target and doses values

In [13]:
# Check for null values again
coverage.isnull().sum()

GROUP                            0
CODE                             0
NAME                             0
YEAR                             0
ANTIGEN                          0
ANTIGEN_DESCRIPTION              0
COVERAGE_CATEGORY                0
COVERAGE_CATEGORY_DESCRIPTION    0
TARGET_NUMBER                    0
DOSES                            0
COVERAGE                         0
dtype: int64

In [14]:
# 2. Incidence data cleaning
incidence = incidence.dropna(subset=['INCIDENCE_RATE'])  # Drop rows with missing incidence rate values

# Check for null values again
incidence.isnull().sum()

GROUP                  0
CODE                   0
NAME                   0
YEAR                   0
DISEASE                0
DISEASE_DESCRIPTION    0
DENOMINATOR            0
INCIDENCE_RATE         0
dtype: int64

In [15]:
# 3. Reported data cleaning
reported = reported.dropna(subset=['CASES'])  # Drop rows with missing reported cases values

# Check for null values again
reported.isnull().sum()

GROUP                  0
CODE                   0
NAME                   0
YEAR                   0
DISEASE                0
DISEASE_DESCRIPTION    0
CASES                  0
dtype: int64

In [16]:
# 4. Vaccine introduction data cleaning
# No missing values
# Check for null values again
vaccine_intro.isnull().sum()

ISO_3_CODE     0
COUNTRYNAME    0
WHO_REGION     0
YEAR           0
DESCRIPTION    0
INTRO          0
dtype: int64

In [17]:
# 5. Vaccine schedule data cleaning
vaccine_schedule = vaccine_schedule.dropna(subset=['TARGETPOP'])  # Drop rows with missing targetpop values
vaccine_schedule = vaccine_schedule.dropna(subset=['GEOAREA'])  # Drop rows with missing geoarea values
vaccine_schedule = vaccine_schedule.dropna(subset=['AGEADMINISTERED'])  # Drop rows with missing ageadministered values
vaccine_schedule = vaccine_schedule.dropna(subset=['SOURCECOMMENT'])  # Drop rows with missing sourcecomment values
# Check for null values again
vaccine_schedule.isnull().sum()

ISO_3_CODE               0
COUNTRYNAME              0
WHO_REGION               0
YEAR                     0
VACCINECODE              0
VACCINE_DESCRIPTION      0
SCHEDULEROUNDS           0
TARGETPOP                0
TARGETPOP_DESCRIPTION    0
GEOAREA                  0
AGEADMINISTERED          0
SOURCECOMMENT            0
dtype: int64

In [18]:
# Dataset shape after cleaning
print("coverage: ", coverage.shape)
print("incidence: ", incidence.shape)
print("reported: ", reported.shape)
print("vaccine_intro: ", vaccine_intro.shape)
print("vaccine_schedule: ", vaccine_schedule.shape)

coverage:  (76473, 11)
incidence:  (61584, 8)
reported:  (65470, 7)
vaccine_intro:  (138320, 6)
vaccine_schedule:  (2266, 12)


## Data Preprocessing Overview

The following data cleaning steps were performed on each dataset to prepare for analysis and ensure data quality:

1. **Coverage Data Cleaning:**
   - Missing country names (`NAME`) were forward filled (`ffill`) to propagate previous valid entries.
   - Rows with missing values in critical columns `TARGET_NUMBER`, `DOSES`, and `COVERAGE` were dropped to ensure completeness of vaccination coverage data.

2. **Incidence Data Cleaning:**
   - Rows missing the `INCIDENCE_RATE` were dropped to maintain accurate disease incidence statistics.

3. **Reported Cases Data Cleaning:**
   - Rows with missing `CASES` values were removed to keep only valid reported disease case entries.

4. **Vaccine Introduction Data Cleaning:**
   - No missing values were present; therefore, no rows were removed or imputed.

5. **Vaccine Schedule Data Cleaning:**
   - Rows missing key schedule information in `TARGETPOP`, `GEOAREA`, `AGEADMINISTERED`, or `SOURCECOMMENT` were dropped to ensure complete vaccination schedule details.

---

## Unit Normalization 

The dataset variables related to vaccination and disease metrics have been reviewed for unit consistency and normalization. Below is the assessment of unit normalization for key fields across tables:

---

#### Coverage Data

- **Coverage (%)**: The `COVERAGE` column represents the percentage of the target population vaccinated. This percentage is consistently reported as a fraction between 0 and 100 for all records.
- **Doses Administered (`DOSES`)**: Represented as absolute counts of doses administered, numeric and consistently used.
- **Target Number (`TARGET_NUMBER`)**: Number of individuals targeted for vaccination, expressed in absolute counts.
- **Coverage Category**: Standardized labels indicating coverage type (e.g., administrative, official).
- **Date (Year)**: Uniform integer year representation.

All numerical units in coverage data are consistently represented and do not require further normalization.

---

#### Incidence Rate Data

- **Incidence Rate**: Expressed as the number of disease cases per specific population unit.
- **Denominator**: Specifies the population basis for incidence calculation (e.g., 'per 1000 live births', 'per 100000 total population').
- The `DENOMINATOR` field clearly defines the unit context, facilitating correct interpretation of incidence values.

Units are explicitly provided, maintaining clarity and normalization across records.

---

#### Reported Cases Data

- **Cases**: Absolute counts of reported disease cases.
- The count data is numeric and uniformly represented.

No inconsistencies detected.

---

#### Vaccine Introduction Data

- Does not include numerical fields needing unit normalization; presence or absence (`INTRO`) is categorical.

---

#### Vaccine Schedule Data

- Contains categorical descriptors for vaccine schedules.
- No numerical unit inconsistencies identified.

---

**Summary**

- All **percentage** and **count** variables are consistently represented in their respective tables.
- **Date fields** use uniform year integers.
- Incidence rate denominators are explicitly indicated per record, eliminating ambiguity.
- No additional unit normalization processing is needed prior to analysis.

This ensures seamless integration and comparability of data across all tables for further exploratory analysis and visualization.

---

# Date Consistency Across Tables

All the datasets in the project consistently use the **Year** field as an integer (`int64` data type), representing the calendar year for the recorded data. This uniform format is present in the following tables:

- **Coverage Data:** `YEAR` column as integer, no missing values.
- **Incidence Rate:** `YEAR` column as integer, no missing values.
- **Reported Cases:** `YEAR` column as integer, no missing values.
- **Vaccine Introduction:** `YEAR` column as integer, no missing values.
- **Vaccine Schedule:** `YEAR` column as integer, no missing values.

The consistent integer format simplifies temporal analysis and querying, enabling straightforward filtering, grouping, and trend visualizations across years without additional date parsing or conversions.

Since the date fields are uniformly formatted as integers, no further formatting or transformation of dates is required at this stage for analysis.

---

## ***3. Database Creation***

Create Database:

In [ ]:
# Create a database connection
conn = sql.connect('vaccination_report.db')
cursor = conn.cursor()

Create Table:

In [ ]:
# 1. Create a table for coverage data
cursor.execute('''
CREATE TABLE IF NOT EXISTS coverage (
    `GROUP` TEXT,
    CODE TEXT,
    NAME TEXT, 
    YEAR INTEGER,
    ANTIGEN TEXT,
    ANTIGEN_DESCRIPTION TEXT,
    COVERAGE_CATEGORY TEXT,
    COVERAGE_CATEGORY_DESCRIPTION TEXT,
    TARGET_NUMBER REAL, 
    DOSES REAL, 
    COVERAGE REAL,
    PRIMARY KEY (CODE)
);''')

# 2. Create a table for incidence data
cursor.execute('''
CREATE TABLE IF NOT EXISTS incidence (
    `GROUP` TEXT,
    CODE TEXT,
    NAME TEXT, 
    YEAR INTEGER,
    DISEASE TEXT,
    DISEASE_DESCRIPTION TEXT,
    DENOMINATOR TEXT,
    INCIDENCE_RATE REAL,
    PRIMARY KEY (CODE),
    FOREIGN KEY (CODE) REFERENCES coverage(CODE)
);''')

# 3. Create a table for reported data
cursor.execute('''
CREATE TABLE IF NOT EXISTS reported_cases (
    `GROUP` TEXT,
    CODE TEXT,
    NAME TEXT, 
    YEAR INTEGER,
    DISEASE TEXT,
    DISEASE_DESCRIPTION TEXT,
    CASES REAL,
    PRIMARY KEY (CODE),
    FOREIGN KEY (CODE) REFERENCES coverage(CODE)
);''')

# 4. Create a table for vaccine introduction data
cursor.execute('''
CREATE TABLE IF NOT EXISTS vaccine_introduction (
    ISO_3_CODE TEXT,
    COUNTRYNAME TEXT,
    WHO_REGION TEXT, 
    YEAR INTEGER,
    INTRO TEXT,
    DESCRIPTION TEXT,
    PRIMARY KEY (ISO_3_CODE),
    FOREIGN KEY (ISO_3_CODE) REFERENCES coverage(CODE)
);''')

# 5. Create a table for vaccine schedule data
cursor.execute('''
CREATE TABLE IF NOT EXISTS vaccine_schedule (
    ISO_3_CODE TEXT,
    COUNTRYNAME TEXT,
    WHO_REGION TEXT, 
    YEAR INTEGER,
    VACCINE_CODE TEXT,
    VACCINE_DESCRIPTION TEXT,
    SCHEDULEROUNDS INTEGER,                
    TARGETPOP TEXT,
    TARGETPOP_DESCRIPTION TEXT,
    GEOAREA TEXT,
    AGEADMINISTERED TEXT,
    SOURCECOMMENT TEXT,
    PRIMARY KEY (ISO_3_CODE),
    FOREIGN KEY (ISO_3_CODE) REFERENCES coverage(CODE)
);''')

# Commit the changes
conn.commit()

print("Tables created successfully")

Tables created successfully


Data Insertion:

In [78]:
# Insert data into coverage table
for _, row in coverage.iterrows():
    cursor.execute("INSERT INTO coverage VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);", tuple(row))

# Commit the changes
conn.commit()

In [79]:
# Insert data into incidence table
for _, row in incidence.iterrows():
    cursor.execute("INSERT INTO incidence VALUES (%s, %s, %s, %s, %s, %s, %s, %s);", tuple(row))

# Commit the changes
conn.commit()
print("Data inserted into incidence table successfully")

Data inserted into incidence table successfully


In [80]:
# Insert data into reported cases table
for _, row in reported.iterrows():
    cursor.execute("INSERT INTO reported_cases VALUES (%s, %s, %s, %s, %s, %s, %s);", tuple(row))

# Commit the changes
conn.commit()
print("Data inserted into reported cases table successfully")

Data inserted into reported cases table successfully


In [81]:
# Insert data into vaccine introduction table
for _, row in vaccine_intro.iterrows():
    cursor.execute("INSERT INTO vaccine_introduction VALUES (%s, %s, %s, %s, %s, %s);", tuple(row))

# Commit the changes
conn.commit()
print("Data inserted into vaccine introduction table successfully")

Data inserted into vaccine introduction table successfully


In [88]:
# Insert data into vaccine schedule table
for _, row in vaccine_schedule.iterrows():
    cursor.execute("INSERT INTO vaccine_schedule VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);", tuple(row))

# Commit the changes
conn.commit()
print("Data inserted into vaccine schedule table successfully")

Data inserted into vaccine schedule table successfully


Display Table:

In [89]:
# Display coverage data from database
coverage_db = pd.read_sql_query('SELECT * FROM coverage;', conn)
coverage_db.head()

C:\Users\Dharun G V\AppData\Local\Temp\ipykernel_20604\1947294817.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  coverage_db = pd.read_sql_query('SELECT * FROM coverage;', conn)


GROUP CODE   NAME  YEAR  ANTIGEN  \
0  COUNTRIES  ABW  Aruba  2023  DIPHCV4   
1  COUNTRIES  ABW  Aruba  2023  DIPHCV5   
2  COUNTRIES  ABW  Aruba  2023  DIPHCV6   
3  COUNTRIES  ABW  Aruba  2023   DTPCV1   
4  COUNTRIES  ABW  Aruba  2023   DTPCV3   

                                 ANTIGEN_DESCRIPTION COVERAGE_CATEGORY  \
0  Diphtheria-containing vaccine, 4th dose (1st b...             ADMIN   
1  Diphtheria-containing vaccine, 5th dose (2nd b...             ADMIN   
2  Diphtheria-containing vaccine, 6th dose (3rd b...             ADMIN   
3                   DTP-containing vaccine, 1st dose             ADMIN   
4                   DTP-containing vaccine, 3rd dose             ADMIN   

  COVERAGE_CATEGORY_DESCRIPTION  TARGET_NUMBER   DOSES  COVERAGE  
0       Administrative coverage         1044.0   945.0     90.52  
1       Administrative coverage         1219.0  1008.0     82.69  
2       Administrative coverage         1468.0  1219.0     83.04  
3       Administrative coverage          945.0   926.0     97.99  
4       Administrative coverage          945.0   907.0     95.98

In [90]:
# Display incidence data from database
incidence_db = pd.read_sql_query('SELECT * FROM incidence', conn)
incidence_db.head()

C:\Users\Dharun G V\AppData\Local\Temp\ipykernel_20604\2132186911.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  incidence_db = pd.read_sql_query('SELECT * FROM incidence', conn)


GROUP CODE   NAME  YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023              CRS   
1  COUNTRIES  ABW  Aruba  2023       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023            MUMPS   
4  COUNTRIES  ABW  Aruba  2023         NTETANUS   

              DISEASE_DESCRIPTION                     DENOMINATOR  \
0     Congenital rubella syndrome          per 10,000 live births   
1                      Diphtheria  per 1,000,000 total population   
2  Invasive meningococcal disease  per 1,000,000 total population   
3                           Mumps  per 1,000,000 total population   
4                Neonatal tetanus           per 1,000 live births   

   INCIDENCE_RATE  
0             0.0  
1             0.0  
2             9.3  
3             0.0  
4             0.0

In [91]:
# Display reported cases data from database
reported_db = pd.read_sql_query('SELECT * FROM reported_cases', conn)
reported_db.head()

C:\Users\Dharun G V\AppData\Local\Temp\ipykernel_20604\1790777522.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  reported_db = pd.read_sql_query('SELECT * FROM reported_cases', conn)


GROUP CODE   NAME  YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023              CRS   
1  COUNTRIES  ABW  Aruba  2023       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023            MUMPS   
4  COUNTRIES  ABW  Aruba  2023         NTETANUS   

              DISEASE_DESCRIPTION  CASES  
0     Congenital rubella syndrome    0.0  
1                      Diphtheria    0.0  
2  Invasive meningococcal disease    1.0  
3                           Mumps    0.0  
4                Neonatal tetanus    0.0

In [92]:
# Display vaccine introduction data from database
vaccine_intro_db = pd.read_sql_query('SELECT * FROM vaccine_introduction', conn)
vaccine_intro_db.head()

C:\Users\Dharun G V\AppData\Local\Temp\ipykernel_20604\1352852003.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vaccine_intro_db = pd.read_sql_query('SELECT * FROM vaccine_introduction', conn)


ISO_3_CODE  COUNTRYNAME WHO_REGION  YEAR  \
0        AFG  Afghanistan       EMRO  2023   
1        AFG  Afghanistan       EMRO  2023   
2        AFG  Afghanistan       EMRO  2023   
3        AFG  Afghanistan       EMRO  2023   
4        AFG  Afghanistan       EMRO  2023   

                                         INTRO DESCRIPTION  
0             aP (acellular pertussis) vaccine          No  
1                          Hepatitis A vaccine          No  
2                          Hepatitis B vaccine         Yes  
3                              HepB birth dose         Yes  
4  Hib (Haemophilus influenzae type B) vaccine         Yes

In [93]:
# Display vaccine schedule data from database
vaccine_schedule_db = pd.read_sql_query('SELECT * FROM vaccine_schedule', conn)
vaccine_schedule_db.head()

C:\Users\Dharun G V\AppData\Local\Temp\ipykernel_20604\4041456260.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vaccine_schedule_db = pd.read_sql_query('SELECT * FROM vaccine_schedule', conn)


ISO_3_CODE COUNTRYNAME WHO_REGION  YEAR VACCINE_CODE  \
0        ABW       Aruba       AMRO  2023      DTAPIPV   
1        ABW       Aruba       AMRO  2023      DTAPIPV   
2        ABW       Aruba       AMRO  2023      DTAPIPV   
3        ABW       Aruba       AMRO  2023   HEPA_ADULT   
4        ABW       Aruba       AMRO  2023   HEPA_ADULT   

            VACCINE_DESCRIPTION  SCHEDULEROUNDS   TARGETPOP  \
0  DTaP-IPV (acellular) vaccine               1  RISKGROUPS   
1  DTaP-IPV (acellular) vaccine               2  RISKGROUPS   
2  DTaP-IPV (acellular) vaccine               3  RISKGROUPS   
3     Adult Hepatitis A vaccine               1  TRAVELLERS   
4     Adult Hepatitis A vaccine               2  TRAVELLERS   

  TARGETPOP_DESCRIPTION   GEOAREA AGEADMINISTERED  \
0         Risk group(s)  NATIONAL     1st contact   
1         Risk group(s)  NATIONAL               0   
2         Risk group(s)  NATIONAL               0   
3            Travellers  NATIONAL     1st contact   
4            Travellers  NATIONAL               0   

                              SOURCECOMMENT  
0  police, firemen, law-enforcement workers  
1  police, firemen, law-enforcement workers  
2  police, firemen, law-enforcement workers  
3                                travellers  
4                                travellers

## Data Visualization in Power BI

## Overview of Key Analytical Questions (Easy Level)

This section outlines the primary questions to be addressed in the vaccination data analysis with a focus on understanding vaccination coverage and its impact on disease incidence. These questions involve demographic, temporal, and geographical factors influencing vaccination outcomes.

### 1. Correlation Between Vaccination Rates and Disease Incidence
- Analyze how increasing vaccination coverage correlates with reductions in the incidence rates of vaccine-preventable diseases.
- Utilize vaccination coverage data alongside disease incidence data over multiple years and regions to quantify this relationship.

### 2. Drop-off Rate Between First Dose and Subsequent Doses
- Measure the percentage decrease in coverage from the initial vaccine dose to subsequent doses (e.g., 1st to 2nd dose, 2nd to 3rd dose).
- Insights can identify where patients are lost in the vaccination schedule and highlight opportunities for intervention.

### 3. Differences in Vaccination Rates by Gender
- No gender column in the dataset.

### 4. Impact of Education on Vaccination Rates
- No education column in the dataset.

### 5. Urban vs Rural Vaccination Coverage
- No urban/rural column in the dataset.

### 6. Trends in Booster Dose Uptake
- Analyze temporal trends in the uptake of booster doses to assess program adherence and effectiveness.
- Increasing booster coverage reflects ongoing commitment to sustained immunity.

### 7. Seasonal Patterns in Vaccination Uptake
- No date column in the dataset.

### 8. Relationship Between Population Density and Vaccination Coverage
- Explore correlations between population density metrics and vaccination coverage.
- High-density zones may pose logistical challenges or opportunities for vaccination drives.

### 10.	Which regions have high disease incidence despite high vaccination rates?

- Certain regions exhibit **persistent high incidence** of vaccine-preventable diseases despite reporting **high vaccination coverage**, possibly due to factors like vaccine efficacy gaps, uneven vaccine distribution, or data reporting discrepancies.

- Identifying these regions is critical for **targeted public health interventions**, addressing underlying causes such as vaccine hesitancy, cold chain issues, or emerging disease strains that reduce vaccine effectiveness.

---


## Medium-Level Analytical Questions Overview

The medium-level analysis involves integrating insights across multiple datasets—vaccine introduction, vaccination coverage, disease incidence, reported cases, and vaccine schedules—to evaluate vaccination effectiveness, program impacts, and geographic and temporal disparities.

### 1. Correlation Between Vaccine Introduction and Disease Reduction
- Analyze timelines of vaccine introductions alongside trends in reported disease cases.
- Determine if the initiation of vaccine programs corresponds to subsequent declines in disease incidence or case counts.
- Requires linking vaccine introduction dates with disease incidence and case data aggregated by country and year.

### 2. Trend Analysis of Disease Cases Before and After Vaccination Campaigns
- Assess historical disease case data across years flanking vaccination program starts.
- Quantify rates of disease reduction post-introduction, controlling for other factors where possible.
- Visualization via time-series plots and interrupted time series analysis enhance interpretation.

### 4. Vaccine Coverage Rates Relative to Target Populations
- Calculate the percentage coverage of vaccines in target populations using administered doses and target population counts.
- Consider variations by antigen and geographic area.
- Enables evaluation of program reach and potential gaps.

### 5. Impact of Vaccination Schedule on Coverage
- Investigate schedule-specific factors such as dose number (initial, booster rounds) and timing on achieved vaccine coverage.
- Understand how scheduling influences coverage drop-off rates and overall immunity levels in populations.

### 6. Disparities in Vaccine Introduction Timelines Across WHO Regions
- Compare the timing of vaccine introductions among WHO regions.
- Identify regions lagging or leading in adopting specific vaccines.
- Inform strategic planning for equitable access.

### 7. Correlation of Coverage with Disease Reduction by Antigen
- Link antigen-specific coverage data with related disease incidence and case trends.
- Explore effectiveness signals and potential vaccine efficacy issues.

### 8. Regions with Low Coverage Despite Vaccine Availability
- Detect areas where vaccines are available but coverage remains suboptimal.
- Highlights challenges in healthcare delivery, acceptance, or supply.

### 10. Geographic Patterns in Disease Prevalence
- Map disease distribution and incidence rates to identify endemic hotspots.
- Integrate with vaccination data to tailor localized public health measures.

---

## Scenario-Based Use Cases Overview 

1. **Identify Regions with Low Vaccination Coverage**  
   Governments seek to pinpoint geographic areas, countries, or subregions exhibiting low rates of vaccine uptake to target interventions and resource allocation more effectively. This involves analyzing vaccination coverage data by region, identifying gaps, and prioritizing areas most in need of support.

2. **Evaluate Effectiveness of a Measles Vaccination Campaign**  
   Public health organizations require assessment of the impact of measles vaccination initiatives conducted in specific countries or regions. This entails comparison of measles incidence and reported case trends before and after the campaign, correlating coverage increases with disease reduction.

3. **Estimate Vaccine Demand for Future Planning**  
   Vaccine manufacturers aim to forecast demand for particular vaccines by analyzing historical coverage trends, target population sizes, and incidence rates. This enables optimizing production, distribution planning, and supply chain management.

4. **Respond to Sudden Influenza Outbreaks**  
   Influenza outbreaks is not found in the dataset.

5. **Assess Polio Incidence in Unvaccinated Populations**  
   Researchers examine the susceptibility and incidence of polio in populations lacking vaccination coverage, providing insights into risk zones and guiding eradication strategies.

6. **Track Global Progress Toward Vaccination Goals**  
   Record of 2023 is not found in the dataset.

7. **Allocate Vaccines to High-Risk Groups**  
   Age group is not found in the dataset.

8. **Detect Disparities in Coverage Across Socioeconomic Groups**  
   Socioeconomic status is not found in the dataset.

9. **Analyze Seasonal Trends in Vaccination Uptake**  
   Authorities study temporal patterns of vaccine administration over months or seasons to optimize timing of vaccination drives and resource allocation.

10. **Compare Effectiveness of Vaccination Strategies**  
    Comparison of different vaccination strategies, such as mass campaigns vs. routine immunization, is not found in the dataset.

---

## Power BI Process

#### Step 1: Prepare the Data Source
- Ensure the `vaccination_report.db` SQLite database is accessible on your system.
- Confirm all cleaned and normalized tables (e.g., coverage_data, incidence_rate, reported_cases, vaccine_intro, vaccine_schedule) are present.

#### Step 2: Connect Power BI to SQLite Database
- Open Power BI Desktop.
- Go to **Home** > **Get Data** > **More...**
- Search for and select **SQLite database** connector.
- Browse and select the `vaccination_report.db` file.
- Load the relevant tables needed for analysis.

#### Step 3: Data Modeling and Relationships
- Inspect the imported tables in the **Model** view.
- Verify or set up relationships according to foreign keys (e.g., linking countries, vaccines, diseases).
- Define calculated columns or measures as needed for analysis, such as vaccination rates or incidence trends.

#### Step 4: Create Visualizations
- Use Power BI's visualization pane to create various charts:
  - **Geographical Heatmaps:** Visualize vaccination coverage and disease incidence by country/region.
  - **Trend Lines/Bar Charts:** Show vaccination trends, disease cases over time.
  - **Scatter Plots:** Correlate vaccination coverage with disease incidence.
  - **KPI Indicators:** Track progress toward targets like 95% coverage.
- Add slicers and filters for dynamic user interaction by year, vaccine type, disease, and region.

#### Step 5: Design Interactive Dashboards
- Combine visuals into report pages emphasizing:
  - Regional disparities.
  - Vaccine effectiveness.
  - High-risk populations.
  - Temporal patterns such as seasonal vaccine uptake.

#### Step 6: Save Power BI Report
- Go to **File** > **Save As**.
- Save the project as `vaccination_report.pbix`.

---

## Conclusion

This vaccination data analysis and visualization project provides a comprehensive framework for understanding global vaccination trends, disease incidence, and vaccine effectiveness. By cleaning, normalizing, and structuring large-scale datasets into a relational SQL database, the project enables robust querying and integration of vaccination coverage, disease cases, vaccine introduction, and schedule information.

The use of Power BI for interactive dashboards facilitates dynamic exploration of vaccination data by policymakers, public health agencies, and researchers. Key insights can be derived to prioritize low-coverage regions, evaluate vaccination campaign outcomes, forecast vaccine demand, and detect disparities across demographic and geographic groups.

Overall, this project equips stakeholders with data-driven tools and visualizations that enhance decision-making in public health strategy, disease prevention, and equity in vaccine access, ultimately contributing to improved population health worldwide.


### ***Hurrah! You have successfully completed your Vaccination Data Analysis and Visualization  Project !!!***